File: src/main/java/xyz/realmath/Rational.java
package xyz.realmath;

import static com.google.common.base.Preconditions.checkArgument;

import com.google.errorprone.annotations.Immutable;
import java.math.BigInteger;

@Immutable
public final class Rational implements Field<Rational> {
  public static final Rational ZERO = Rational.valueOf(0);
  public static final Rational ONE = Rational.valueOf(1);

  private final BigInteger numer;
  private final BigInteger denom;

  private Rational(BigInteger numer, BigInteger denom) {
    checkArgument(!denom.equals(BigInteger.ZERO), "denominator must not be zero");
    BigInteger oldNumer = numer;

    numer = numer.divide(oldNumer.gcd(denom));
    denom = denom.divide(oldNumer.gcd(denom));

    if (denom.signum() == -1) {
      numer = numer.negate();
      denom = denom.negate();
    }

    this.numer = numer;
    this.denom = denom;
  }

  private Rational(int i) {
    this(BigInteger.valueOf(i), BigInteger.ONE);
  }

  public static Rational valueOf(int i) {
    return new Rational(i);
  }

  public static Rational factorial(int n) {
    Rational retVal = Rational.ONE;
    for (int i = 1; i <= n; i++) {
      retVal = retVal.multiply(i);
    }
    return retVal;
  }

  @Override
  public String toString() {
    if (denom.equals(BigInteger.ONE)) {
      return numer.toString();
    } else {
      return numer.toString() + ":" + denom.toString();
    }
  }

  @Override
  public Rational add(Rational by) {
    BigInteger newDenom = denom.multiply(by.denom);
    BigInteger newNumer = numer.multiply(by.denom);

    newNumer = newNumer.add(by.numer.multiply(denom));

    return new Rational(newNumer, newDenom);
  }

  @Override
  public Rational add(int i) {
    return this.add(Rational.valueOf(i));
  }

  @Override
  public Rational multiply(Rational by) {
    return new Rational(numer.multiply(by.numer), denom.multiply(by.denom));
  }

  @Override
  public Rational multiply(int i) {
    return this.multiply(Rational.valueOf(i));
  }

  @Override
  public Rational divide(int i) {
    return this.divide(Rational.valueOf(i));
  }

  @Override
  public Rational negate() {
    return new Rational(numer.negate(), denom);
  }

  @Override
  public Rational reciprocal() {
    return new Rational(denom, numer);
  }

  /** Power. */
  public Rational pow(int n) {
    if (n <= 0) {
      checkArgument(!this.equals(Rational.ZERO), "undefined");
    }
    if (n < 0) {
      return pow(-n).reciprocal();
    }
    Rational retVal = Rational.ONE;
    for (int i = 0; i < n; i++) {
      retVal = retVal.multiply(this);
    }
    return retVal;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) {
      return true;
    }
    if (obj == null || getClass() != obj.getClass()) {
      return false;
    }
    Rational rational = (Rational) obj;
    return java.util.Objects.equals(numer, rational.numer)
        && java.util.Objects.equals(denom, rational.denom);
  }

  @Override
  public int hashCode() {
    return java.util.Objects.hash(numer, denom);
  }
}

File: src/main/kotlin/xyz/realmath/Rational.kt
package xyz.realmath

import java.math.BigInteger

class Rational private constructor(numer: BigInteger, denom: BigInteger) : Field<Rational> {

    private val numer: BigInteger
    private val denom: BigInteger

    init {
        check(denom != BigInteger.ZERO) { "denominator must not be zero" }

        var numer1 = numer.divide(numer.gcd(denom))
        var denom1 = denom.divide(numer.gcd(denom))

        if (denom1.signum() == -1) {
            numer1 = -numer1
            denom1 = -denom1
        }

        this.numer = numer1
        this.denom = denom1
    }

    constructor(i: Int) : this(BigInteger.valueOf(i.toLong()), BigInteger.ONE)

    override fun toString(): String {
        return if (denom == BigInteger.ONE) {
            numer.toString()
        } else {
            numer.toString() + ":" + denom.toString()
        }
    }

    override operator fun plus(t: Rational): Rational {
        val newDenom = denom * t.denom
        var newNumer = numer * t.denom

        newNumer += t.numer * denom

        return Rational(newNumer, newDenom)
    }

    override operator fun plus(i: Int): Rational {
        return this + Rational(i)
    }

    override fun zero() = ZERO

    override fun one() = ONE

    override operator fun times(multiplier: Rational): Rational {
        return Rational(numer * multiplier.numer, denom * multiplier.denom)
    }

    override operator fun times(i: Int): Rational {
        return this * Rational(i)
    }

    override operator fun div(i: Int): Rational {
        return this / Rational(i)
    }

    override fun reciprocal(): Rational {
        check(!isZero) { "zero reciprocal" }
        return Rational(denom, numer)
    }

    override operator fun unaryMinus(): Rational {
        return Rational(-numer, denom)
    }

    /** Power.  */
    fun pow(n: Int): Rational {
        if (n <= 0) {
            check(this != Rational.ZERO) { "undefined" }
        }
        if (n < 0) {
            return Rational.ONE / pow(-n)
        }
        var retVal = Rational.ONE
        for (i in 0 until n) {
            retVal *= this
        }
        return retVal
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Rational) return false

        if (numer != other.numer) return false
        if (denom != other.denom) return false

        return true
    }

    override fun hashCode(): Int {
        var result = numer.hashCode()
        result = 31 * result + denom.hashCode()
        return result
    }

    companion object {
        val ZERO = Rational(0)
        val ONE = Rational(1)

        fun factorial(n: Int): Rational {
            var retVal = Rational.ONE
            for (i in 1..n) {
                retVal *= i
            }
            return retVal
        }
    }
}
